ddl/partition.go

```diff
@@ -2164,6 +2188,316 @@ func (w *worker) onExchangeTablePartition(d *ddlCtx, t *meta.Meta, job *model.Jo
+	schemaID := job.SchemaID
+	tblInfo, err := GetTableInfoAndCancelFaultJob(t, job, schemaID)
+	if err != nil {
+		return nil, nil, nil, nil, nil, errors.Trace(err)
+	}
+	partInfo := &model.PartitionInfo{}
+	var partNames []model.CIStr
+	err = job.DecodeArgs(&partNames, &partInfo)
+	if err != nil {
+		job.State = model.JobStateCancelled
+		return nil, nil, nil, nil, nil, errors.Trace(err)
+	}
+	addingDefs := tblInfo.Partition.AddingDefinitions
+	droppingDefs := tblInfo.Partition.DroppingDefinitions
+	if len(addingDefs) == 0 {
+		addingDefs = []model.PartitionDefinition{}
+	}
+	if len(droppingDefs) == 0 {
+		droppingDefs = []model.PartitionDefinition{}
+	}
+	return tblInfo, partNames, partInfo, droppingDefs, addingDefs, nil
+}
+
+func (w *worker) onReorganizePartition(d *ddlCtx, t *meta.Meta, job *model.Job) (ver int64, _ error) {
+	// Handle the rolling back job
+	if job.IsRollingback() {
+		ver, err := w.onDropTablePartition(d, t, job)
+		if err != nil {
+			return ver, errors.Trace(err)
+		}
+		return ver, nil
+	}
+
+	tblInfo, partNamesCIStr, partInfo, _, addingDefinitions, err := checkReorgPartition(t, job)
+	if err != nil {
+		return ver, err
+	}
+	partNames := make([]string, len(partNamesCIStr))
+	for i := range partNamesCIStr {
+		partNames[i] = partNamesCIStr[i].L
+	}
+
+	// In order to skip maintaining the state check in partitionDefinition, TiDB use dropping/addingDefinition instead of state field.
+	// So here using `job.SchemaState` to judge what the stage of this job is.
+	originalState := job.SchemaState
+	switch job.SchemaState {
+	case model.StateNone:
+		// job.SchemaState == model.StateNone means the job is in the initial state of reorg partition.
+		// Here should use partInfo from job directly and do some check action.
+		// In case there was a race for queueing different schema changes on the same
+		// table and the checks was not done on the current schema version.
+		// The partInfo may have been checked against an older schema version for example.
+		// If the check is done here, it does not need to be repeated, since no other
+		// DDL on the same table can be run concurrently.
+		err = checkAddPartitionTooManyPartitions(uint64(len(tblInfo.Partition.Definitions) +
+			len(partInfo.Definitions) -
+			len(partNames)))
+		if err != nil {
+			job.State = model.JobStateCancelled
+			return ver, errors.Trace(err)
+		}
+
+		err = checkReorgPartitionNames(tblInfo.Partition, partNamesCIStr, partInfo)
+		if err != nil {
+			job.State = model.JobStateCancelled
+			return ver, errors.Trace(err)
+		}
+
+		// Re-check that the dropped/added partitions are compatible with current definition
+		firstPartIdx, lastPartIdx, idMap, err := getReplacedPartitionIDs(partNamesCIStr, tblInfo.Partition)
+		if err != nil {
+			return ver, err
+		}
+		var sctx sessionctx.Context
```

Comment:
I'm curious if the check `sess == nil` is needed. After all, we used it almost everywhere.

PTAL @wjhuang2016 @tangenta

Reply:
No need to check it. It can't be null